test/unit/gateways/secure_co_test.rb

```diff
@@ -0,0 +1,256 @@
+    assert_success response
+  end
+
+  def test_failed_verify
+    @gateway.expects(:ssl_post).returns(failed_authorize_response)
+    response = @gateway.verify(@credit_card, @options)
+    assert_failure response
+  end
+
+  def test_scrub
+    assert @gateway.supports_scrubbing?
+    assert_equal @gateway.scrub(pre_scrubbed), post_scrubbed
+  end
+
+  def test_optional_fields
+    stub_comms do
+      @gateway.purchase(2000, @credit_card, @options.merge(:currency => 'CAD'))
+    end.check_request do |endpoint, data, headers|
+      assert_match '<requested-amount currency="CAD">20.00</requested-amount>', data
+      assert_match '<account-number>4111111111111111</account-number>', data
+      assert_match '<card-security-code>123</card-security-code>', data
+      assert_match %r{<request-id>[0-9a-f]{32}<\/request-id>}, data
+    end.respond_with(successful_purchase_response)
+
+    stub_comms do
+      @gateway.purchase(2000, @credit_card, @options.merge(
+        ip: '127.0.0.1',
+        request_id: '00000000-0000-0000-0000-000000000000',
+        order_id: 'ORDER1234',
+        custom_fields: [["field1", "value1"], [:field2, "value2"], ["field three", 3]],
+        email: 'test.email@secureco.co',
+      ))
+    end.check_request do |endpoint, data, headers|
+      assert_match '<ip-address>127.0.0.1</ip-address>', data
+      assert_match '<request-id>00000000-0000-0000-0000-000000000000</request-id>', data
+      assert_match '<order-number>ORDER1234</order-number>', data
+      assert_match '<custom-field field-name="field1" field-value="value1"/>', data
+      assert_match '<custom-field field-name="field2" field-value="value2"/>', data
+      assert_match '<custom-field field-name="field three" field-value="3"/>', data
+      assert_match '<email>test.email@secureco.co</email>', data
+    end.respond_with(successful_purchase_response)
+
+    stub_comms do
+      @gateway.purchase(2000, @credit_card, @options.merge(
+        custom_fields: {"field1" => "value1", :field2 => "value2", "field three" => 3},
+      ))
+    end.check_request do |endpoint, data, headers|
+      refute_match %r{<ip-address\b}, data
+      refute_match %r{<order-number\b}, data
+      assert_match '<custom-field field-name="field1" field-value="value1"/>', data
+      assert_match '<custom-field field-name="field2" field-value="value2"/>', data
+      assert_match '<custom-field field-name="field three" field-value="3"/>', data
+      refute_match %r{<email\b}, data
+    end.respond_with(successful_purchase_response)
+  end
+
+  def test_dependent_order_types
+    @gateway.expects(:ssl_post).twice.returns(successful_purchase_response)
+    purchase_response = @gateway.purchase(@amount, @credit_card, @options)
+    stub_comms do
+      @gateway.refund(@amount, purchase_response.authorization, @options)
+    end.check_request do |endpoint, data, headers|
+      assert_match '<transaction-type>refund-purchase</transaction-type>', data
```

Comment:
I wasn't able to recreate the error, but I strongly suspected it was something to do with the way I was mixing two ways of http request stubbing.

So I refactored those tests and discovered that half of them weren't even being run. I don't even know how that's possible without raising errors.

I just pushed my changes.